src/app/src/test/kotlin/org/vaccineimpact/orderlyweb/tests/unit_tests/controllers/web/WorkflowRunControllerTests.kt

```diff
@@ -97,10 +97,127 @@ class WorkflowRunControllerTests
             on { getWorkflowRunDetails("fakeKey") } doThrow UnknownObjectError("key", "workflow")
         }
 
-        val sut = WorkflowRunController(mockContext, mockRepo)
+        val sut = WorkflowRunController(mockContext, mockRepo, mock())
 
         Assertions.assertThatThrownBy { sut.getWorkflowRunDetails() }
-                .isInstanceOf(UnknownObjectError::class.java)
-                .hasMessageContaining("Unknown workflow : 'key'")
+            .isInstanceOf(UnknownObjectError::class.java)
+            .hasMessageContaining("Unknown workflow : 'key'")
     }
+
+    @Test
+    fun `can deserialize a workflow description`()
+    {
+        val json = """
+            {
+              "name": "workflow1",
+              "instances": {
+                "database1": "instance1"
+              },
+              "changelog": {
+                "message": "message1",
+                "type": "type1"
+              },
+              "git_branch": "branch1",
+              "git_commit": "commit1",
+              "reports": [
+                {
+                  "name": "report1",
+                  "params": {
+                    "param1": "value1"
+                  }
+                },
+                {
+                  "name": "report2",
+                  "params": {
+                    "param2": "value2"
+                  }
+                }
+              ]
+            }
+        """.trimIndent()
+
+        var workflowRunRequest = Serializer.instance.gson.fromJson(json, WorkflowRunRequest::class.java)
+        assertThat(workflowRunRequest).isEqualTo(getWorkflowRunRequestExample())
+    }
+
+    @Test
+    fun `can run a workflow`()
+    {
+        val workflowRunRequest = getWorkflowRunRequestExample()
+
+        val context = mock<ActionContext> {
+            on { getRequestBody() } doReturn Serializer.instance.gson.toJson(workflowRunRequest)
+            on { userProfile } doReturn CommonProfile().apply { id = "test@user.com" }
+        }
+
+        val mockAPIResponseText = """{"data": {"workflow_key": "workflow_key1", "reports": ["report_key1"]}}"""
+
+        val mockAPIResponse = OrderlyServerResponse(mockAPIResponseText, 200)
+
+        val apiClient: OrderlyServerAPI = mock {
+            on { post(any(), any<String>(), any()) } doReturn mockAPIResponse
+        }
+
+        val repo = mock<WorkflowRunRepository>()
+        val sut = WorkflowRunController(context, repo, apiClient)
+        val result = sut.createWorkflowRun()
+
+        verify(apiClient).post(
+            "/v1/workflow/run/",
+            Serializer.instance.gson.toJson(
+                mapOf(
+                    "changelog" to mapOf(
```

Comment:
Could maybe do with a test that null changelog and refs are not included

Reply:
Done - thanks